File: 0070-climbing-stairs/0070-climbing-stairs.java
class Solution {
    public int climbStairs(int n) {
        // if(n == 1) return 1;

        // int[] dp = new int[n + 1];
        // dp[0] = 1;
        // dp[1] = 1;

        // for(int i = 2 ; i <= n ; i++)
        // {
        //     dp[i] = dp[i-1] + dp[i-2];
        // }
        // return dp[n];

        // decrease space complexity to O(1)

        if(n == 1) return 1;

        int a = 1, b = 1;

        for(int i = 2 ; i <= n ; i++)
        {
            int temp = a + b;
            a = b;
            b = temp;
        }
        return b;
    }
}

File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
    vector<int> memo;
public:
    int climbStairs(int n) {
        memo.resize(n + 1);
        return dfs(n);
    }

    int dfs(int i){
        if(i <= 1) return 1;
        int& res = memo[i];
        if(res) return res;
        return res = dfs(i - 1) + dfs(i - 2);
    }
};